libimage/runtime.go

```diff
@@ -316,19 +316,36 @@ func (r *Runtime) LookupImage(name string, options *LookupImageOptions) (*Image,
 		return img, name, err
 	}
 
-	return r.lookupImageInDigestsAndRepoTags(name, options)
+	return r.lookupImageInRepoTags(name, namedName, options)
 }
 
 // lookupImageInLocalStorage looks up the specified candidate for name in the
 // storage and checks whether it's matching the system context.
 func (r *Runtime) lookupImageInLocalStorage(name, candidate string, options *LookupImageOptions) (*Image, error) {
 	logrus.Debugf("Trying %q ...", candidate)
+
+	// First, try store.Image() which will properly work on IDs (but not on
+	// digests).
 	img, err := r.store.Image(candidate)
 	if err != nil && !errors.Is(err, storage.ErrImageUnknown) {
 		return nil, err
 	}
 	if img == nil {
-		return nil, nil
+		// Second, try parsing the candidate into a storage reference
+		// which will work with digests.  However, we must reparse the
+		// reference another time below since an ordinary image may
+		// have been referenced via its parent (manifest list) digest.
```

Comment:
If I understand this correctly, this is referring to a problem with `storageTransport.Transport.GetStoreImage`.

Without that problem, `storageReference.resolveImage` _already_ intends to choose the right per-platform variant; it wouldn’t typically return a multi-arch reference at all (unless the image is manifest-only I guess?).

Do I get that right?

---

The problem with `GetStoreImage` is two-fold:
- It doesn’t have a `SystemContext` parameter, i.e. it calls `resolveImage` possibly choosing the wrong image
- It doesn’t _start_ with `resolveImage`; it first does a dumb lookup via `store.Image(referenceNamed.String())`. I’m not 100% sure why that exists at all.

If that’s what this is about, I suggest a new `c/image/storage.GetStoreImage` function fixing these two things would help.

Reply:
IOW, I think involving `GetStoreImage` fixes lookups by digest, but the cost is, right now, breaking `options.{Architecture,OS,Variant}` — and maybe `options.{lookupManifest,IsManifestList}`, I’m not sure about that part.